relay-server/src/actors/upstream.rs

```diff
@@ -223,14 +245,27 @@ struct UpstreamRequest {
 }
 
 pub struct UpstreamRelay {
+    /// backoff policy for the registration messages
     backoff: RetryBackoff,
+    /// backoff policy for the network outage message
+    network_status_backoff: RetryBackoff,
+    /// from this instant forward we only got network errors on all our http requests
+    /// (any request that is sent without causing a network error resets this back to None)
     first_error: Option<Instant>,
     config: Arc<Config>,
     auth_state: AuthState,
     max_inflight_requests: usize,
     num_inflight_requests: usize,
+    /// queue holding the high priority messages to be sent
     high_prio_requests: VecDeque<UpstreamRequest>,
+    /// queue holding the low priority messages to be sent
     low_prio_requests: VecDeque<UpstreamRequest>,
+    /// Set to true when the Upstream explicitly sends a Stop authentication response
+    /// or there is a client error and there is no point to try to authenticate again.
+    authentication_denied: bool,
```

Comment:
Can we merge this with `auth_state`? I feel like this is duplicate and could be replaced by `AuthState::Denied`, which integrates well into our state machine. 

Same for the `is_waiting_to_authenticate` flag, do we in fact need this? It seems just like a shorthand for certain auth states.

The full state machine should be:

- on success: `Unknown -> Authenticating -> Authenticated <-> Reauthenticating`
- on error: `-> Authenticating -> ...`
- on deny: `-> Denied (terminal)`

The difference here is that `Authenticating` indicates not being authenticated, and `Reauthenticating` indicates being authenticated.

Reply:
good point about `Denied`.

`is_waiting_to_authenticate ` is orthogonal with the other states, it can occur in `Unknown`, `Authenticated` and `Error`. It signifies that there is a network outage and immediately after it is resolved we want to do an authentication. Basically this is set when we try to authenticate but find out we are in an outage and we have to postpone our authentication until the network is up again.**Let's talk about this one.**